Add the sriov network operator ci project

abdallahyas opened the pull request:
Add the sriov_network_operator project to test the e2e tests of
the sriov-network-operator on nvidia hardware.

AbdYsn commented:
Patch is tested and should be working, but it needs wrapping up (docs and nits here and there)

Mmduh-483 commented:
Can you split the commits as follow:
- Updating prepare_setup.sh script
- Install and deploy kind
- Background script to map a kind node with PF and its VFs
- Add sriov network operator CI
- Update README

abdallahyas committed:
Add skapoe packages to the prepare_setup.sh script

Added the skapoe binary build to the `prepare_setup.sh` in
preparation to use them in future CIs.

abdallahyas committed:
Add a bash function wrappers for `run_kind_ci.sh`

Added two functions to deploy and undeploy a kind a cluster using
`run_kind_ci.sh`. This is needed to use specific phases of the script
since not all of them is needed.

abdallahyas committed:
Add vf-switcher service deployment functions

Added the vf-switcher service configurations and deployment functions
to the project , in preparation to use them in future CI.

abdallahyas committed:
Add sriov network operator project

Add the sriov_network_operator project to test the e2e tests of
the sriov-network-operator on nvidia hardware.

abdallahyas committed:
Add the README of the sriov_network_operator project

Added a simple README to the sriov_network_operator project to
describe briefly what it does.

Mmduh-483 commented on prepare_setup.sh:
```diff
@@ -7,6 +7,8 @@ go_dir=/usr/local
 helm_version="v3.4.0"
 helm_package_dir_name='helm'
 
+kind_version="v0.10.0"
```
Use github api to get the latest release `https://api.github.com/repos/kubernetes-sigs/kind/releases/latest` , it returns a json, `result['name']`

abdallahyas replied:
IMO, pinning it to a certain version would be best to not be surprised of different behavior when preparing new nodes, though we will need to figure out a way to install latest stable Kubenetes in the node image.

Mmduh-483 replied:
> IMO, pinning it to a certain version would be best to not be surprised of different behavior when preparing new nodes, though we will need to figure out a way to install latest stable Kubenetes in the node image.

In k8s ci we are using the latest official release to find if there will be any surprises later and I belive it should be the same for kind

abdallahyas replied:
That is true for deployment components, but kind is a tool we use to deploy those components, so i think pinning kind and building the node image with latest stable Kubernetes is the correct way of doing it. since also latest kind does not necessarily deploy the latest Kubernetes.

Mmduh-483 commented on prepare_setup.sh:
```diff
@@ -270,6 +272,43 @@ packages_install(){
+
+kind_install(){
+    wget https://github.com/kubernetes-sigs/kind/releases/download/"$kind_version"/kind-linux-amd64
+    mv kind-linux-amd64 /usr/local/bin/kind
```
use `wget -O` instead of `mv`

abdallahyas replied:
Done

Mmduh-483 commented on common/common_functions.sh:
```diff
@@ -817,6 +817,120 @@ function deploy_k8s_bare {
+    fi
+
+    for worker_container in $worker_containers; do
+        sudo docker exec -it "$worker_container" mount -o remount,rw /sys
```
Why is this needed?

abdallahyas replied:
This is need to be able to create VFs, since to create them, write access to the sriov_numvfs is needed. and kind by default remount the sys fs to be read only. i think they do so to not have many container (nodes) writing to the same file.

Mmduh-483 replied:
Is it configurable in kind to mount it with write permission?

abdallahyas replied:
no, it is hard coded, see https://github.com/kubernetes-sigs/kind/blob/b6bc112522651d98c81823df56b7afa511459a3b/images/base/files/usr/local/bin/entrypoint#L147

ykulazhenkov replied:
I think it is possible to use more specific paths in kind config, then it should work, eg
```
   extraMounts:
    - hostPath: /sys/class/net
      containerPath: /sys/class/net
    - hostPath: /sys/devices
      containerPath: /sys/devices
```

abdallahyas replied:
@ykulazhenkov  unfortunately this will not work also, since this basically mount the directory inside the container, then the entrypoint will do the remount to read only after the container is started.

P.S. I did try this the first time i faced the issue, but that did not work.

Mmduh-483 commented on common/common_functions.sh:
```diff
@@ -960,3 +1074,16 @@ function load_core_drivers {
+    fi
+
+    echo "upload $image to kind...."
+    kind load docker-image $image
```
You didn't specify the cluster name

abdallahyas replied:
added it

Mmduh-483 commented on deploy/vf-switcher.service:
```diff
@@ -0,0 +1,12 @@
+[Unit]
```
Why need the service, vf switcher is a background job it is not needed at boot

abdallahyas replied:
using services, it would be easier to start, stop, and collect logs across scripts. It would also be easier to debug in case of failures.

Mmduh-483 commented on sriov_network_operator/sriov_network_operator_ci_start.sh:
```diff
@@ -0,0 +1,144 @@
+
+    label_node
+
+    make deploy-setup-k8s
```
you didn't specify the operator `SRIOV_NETWORK_OPERATOR_IMAGE ` and daemon `SRIOV_NETWORK_CONFIG_DAEMON_IMAGE ` images and the cluster type `CLUSTER_TYPE`

sriov operator uses latest master if they are not speicfied and cluster type to openshift

abdallahyas replied:
see [Makefile](https://github.com/k8snetworkplumbingwg/sriov-network-operator/blob/master/Makefile#L195-L200) for basic env variables, and see [vars](https://github.com/Mellanox/kubernetes-ci/pull/105/files?file-filters%5B%5D=.sh#diff-1da590bb51c16903726785c833421c5daa35025cdc36c923aca64459aae22f1cR17-R21) for the image variables

Mmduh-483 replied:
I see you set the env vars for images, what about the cluster type?

abdallahyas replied:
it is set in the Makefile of the sriov-network operator

Mmduh-483 commented on sriov_network_operator/README.md:
```diff
@@ -0,0 +1,3 @@
+# SRIOV NETWORK OPERATOR project
```
Please add more details about the scenario tested and the passed parameters

abdallahyas replied:
Done

ykulazhenkov commented on common/common_functions.sh:
```diff
@@ -817,6 +817,120 @@ function deploy_k8s_bare {
+
+    deploy_kind_cluster "$cluster_name" "$workers_number"
+    let status=status+$?
+    if [ "$status" != 0 ]; then
```
nit: you mixing "[" and "[[" syntax for the same checks. It is better to use consistent style

ykulazhenkov commented on common/common_functions.sh:
```diff
@@ -817,6 +817,120 @@ function deploy_k8s_bare {
+    local workers_number="$2"
+
+    deploy_kind_cluster "$cluster_name" "$workers_number"
+    let status=status+$?
```
It looks like here is lack of `status` initialization `local status=0`

Should be `let status=status+$?` replaced with `let status=$status+$`?

ykulazhenkov commented on common/common_functions.sh:
```diff
@@ -817,6 +817,120 @@ function deploy_k8s_bare {
+    fi
+
+    deploy_multus
+    let status=status+$?
```
`let status=status+$?` >> `let status=$status+$`

ykulazhenkov commented on common/common_functions.sh:
```diff
@@ -1139,13 +1139,14 @@ function load_core_drivers {
 
     echo "upload $image to kind...."
-    kind load docker-image $image
+    kind load docker-image $image --name $cluster_name
```
It is better to move addition of cluster name to the commit where you add upload_image_to_kind function

ykulazhenkov commented on deploy/vf-netns-switcher.sh:
```diff
@@ -0,0 +1,243 @@
+    for key in $(yq r "$conf_file" | cut -d ":" -f 1);do
+        local corrected_key="$(tr '-' '_' <<< "$key")"
+
+	eval ${corrected_key}="$(yq r $conf_file $key)"
```
confusing indentation, tab used instead of spaces

ykulazhenkov commented on deploy/vf-netns-switcher.sh:
```diff
@@ -0,0 +1,243 @@
+    if [[ -z "$(ip netns | grep $worker_netns)" ]];then
+        local pid="$(docker inspect -f '{{.State.Pid}}' $worker_netns)"
+
+	if [[ -z "pid" ]];then
```
"pid" > "$pid" ?

ykulazhenkov commented on deploy/vf-netns-switcher.sh:
```diff
@@ -0,0 +1,243 @@
+    local vf_name="$1"
+    local worker_netns="${2:-$netns}"
+
+    if [[ -z "ip l show $vf_name" ]];then
```
subshell $() missing

ykulazhenkov commented on common/common_functions.sh:
```diff
@@ -817,6 +817,182 @@ function deploy_k8s_bare {
+    yaml_write 'netns' "$worker_name" "$conf_file"
+    let status=$status+$?
+
+    rm -rf /var/run/netns/"$worker_name"
```
I think you should move this to somewhere else. Function with name "prepare_vf_switcher_confs" should not remove symlink for container's ns.

ykulazhenkov commented on prepare_setup.sh:
```diff
@@ -270,6 +272,43 @@ packages_install(){
+
+kind_install(){
+    wget https://github.com/kubernetes-sigs/kind/releases/download/"$kind_version"/kind-linux-amd64\
+        -O /usr/local/bin/kind
```
Please, move "mv to -O" fix to the commit where you add kind_install function

ykulazhenkov commented on deploy/vf-netns-switcher.sh:
```diff
@@ -0,0 +1,243 @@
+
+   --help | -h)
+      echo "
+vf-netns-switcher.sh --netns <> --vf-driver <> --vf-count <>:
```
I can't see usage of `--vf-driver <> --vf-count <>`

ykulazhenkov commented on deploy/vf-netns-switcher.sh:
```diff
@@ -0,0 +1,243 @@
+    local conf_file="$1"
+
+    for key in $(yq r "$conf_file" | cut -d ":" -f 1);do
+        local corrected_key="$(tr '-' '_' <<< "$key")"
```
Maybe we can drop this? It look like we have no keys with "-" in the config at the moment.

ykulazhenkov commented on deploy/vf-netns-switcher.sh:
```diff
@@ -0,0 +1,243 @@
+        echo ""
+        return 1
+    fi
+    grep PCI_SLOT_NAME /sys/class/net/"${interface_name}"/device/uevent | cut -d '=' -f 2
```
Maybe it will be better to use something like this?
```basename $(readlink /sys/class/net/enp5s0f0/device)```

ykulazhenkov commented on deploy/vf-netns-switcher.sh:
```diff
@@ -0,0 +1,243 @@
+    grep PCI_SLOT_NAME /sys/class/net/"${interface_name}"/device/uevent | cut -d '=' -f 2
+}
+
+check_driver(){
```
unused function

ykulazhenkov commented on deploy/vf-netns-switcher.sh:
```diff
@@ -0,0 +1,243 @@
+
+    if [[ -z "$(ip l show $interface_name)" ]];then
+        if [[ -n "$(docker exec -t ${worker_netns} ip l show $interface_name)" ]];then
+             docker exec ${worker_netns} grep PCI_SLOT_NAME /sys/class/net/"${interface_name}"/device/uevent | cut -d '=' -f 2
```
Maybe it will be better to use here `ip netns e <ns> <cmd>` instead of docker exec?

ykulazhenkov commented on sriov_network_operator/sriov_network_operator_ci_test.sh:
```diff
@@ -0,0 +1,43 @@
+export TIMEOUT=${TIMEOUT:-300}
+
+export POLL_INTERVAL=${POLL_INTERVAL:-10}
+export NETWORK=${NETWORK:-'192.168'}
```
All these variables are needed here?

ykulazhenkov commented on deploy/vf-netns-switcher.sh:
```diff
@@ -0,0 +1,234 @@
+    if [[ ! -e /var/run/netns/$worker_netns ]];then
+        local pid="$(docker inspect -f '{{.State.Pid}}' $worker_netns)"
+
+	if [[ -z "$pid" ]];then
```
nit: in this line tab used instead of spaces for indentation

abdallahyas replied:
thanks for the review and sorry for these small nits, i will investigate if we can have some kind of CI that would do customized linting for bash.

ykulazhenkov commented on deploy/vf-netns-switcher.sh:
```diff
@@ -0,0 +1,234 @@
+    if [[ -z "$(ip l show $interface_name)" ]];then
+        if [[ -n "$(docker exec -t ${worker_netns} ip l show $interface_name)" ]];then
+             ip netns exec ${worker_netns} basename $(readlink /sys/class/net/${interface_name}/device)
+	     return 0
```
nit: tab + spaces used for this line indentation

abdallahyas replied:
removed

ykulazhenkov commented on common/common_functions.sh:
```diff
@@ -817,6 +817,135 @@ function deploy_k8s_bare {
+
+    local status=0
+
+    ./run_kind_ci.sh --project "$project" --num-workers "$workers_number"\
```
It looks like this call can fail after this PR  [1] will be merged 
I think it is better to discuss with @Mmduh-483 which PR should be merged first.
[1] https://github.com/Mellanox/kubernetes-ci/pull/110/files

abdallahyas replied:
you are correct, phases that should not run will run now. @Mmduh-483 WDYT?

Mmduh-483 replied:
I think we can add a flag `--phases` to run specific phases and calling this flag will ignore `--skip-phases` for simplicty

Mmduh-483 replied:
Check PR #112

abdallahyas replied:
updated the PR, can you have a look

Mmduh-483 replied:
You need to run `prepare-ci-environment` phase also. `deploy-kind/create-cluster` sub-phase renders the config file then try to move it to the artifacts dir that created in `prepare-ci-environment`

abdallahyas replied:
@Mmduh-483 I do that manually in the start script. see https://github.com/Mellanox/kubernetes-ci/pull/105/files#diff-1da590bb51c16903726785c833421c5daa35025cdc36c923aca64459aae22f1cR110

e0ne reviewed (approved):
LGTM
